kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -0,0 +1,331 @@
+import com.fasterxml.jackson.jaxrs.base.JsonParseExceptionMapper;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import io.confluent.kafkarest.common.CompletableFutures;
+import io.confluent.kafkarest.exceptions.StatusCodeException;
+import io.confluent.kafkarest.exceptions.v3.ErrorResponse;
+import io.confluent.kafkarest.exceptions.v3.V3ExceptionMapper;
+import io.confluent.rest.entities.ErrorMessage;
+import io.confluent.rest.exceptions.KafkaExceptionMapper;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import javax.ws.rs.ext.ExceptionMapper;
+import org.glassfish.jersey.server.ChunkedOutput;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helper class to allow transforming a stream of requests into a stream of responses online.
+ *
+ * <p>Usage:
+ *
+ * <pre>{@code
+ * StreamingResponse.from(myStreamOfRequests) // e.g. a MappingIterator request body
+ *     .compose(request -> computeResponse(request))
+ *     .resume(asyncResponse);
+ * }</pre>
+ */
+// CHECKSTYLE:OFF:ClassDataAbstractionCoupling
+public abstract class StreamingResponse<T> {
+
+  private static final Logger log = LoggerFactory.getLogger(StreamingResponse.class);
+
+  private static final CompositeErrorMapper EXCEPTION_MAPPER =
+      new CompositeErrorMapper.Builder()
+          .putMapper(
+              JsonMappingException.class,
+              new JsonMappingExceptionMapper(),
+              response -> Status.BAD_REQUEST.getStatusCode(),
+              response -> (String) response.getEntity())
+          .putMapper(
+              JsonParseException.class,
+              new JsonParseExceptionMapper(),
+              response -> Status.BAD_REQUEST.getStatusCode(),
+              response -> (String) response.getEntity())
+          .putMapper(
+              StatusCodeException.class,
+              new V3ExceptionMapper(),
+              response -> ((ErrorResponse) response.getEntity()).getErrorCode(),
+              response -> ((ErrorResponse) response.getEntity()).getMessage())
+          .setDefaultMapper(
+              new KafkaExceptionMapper(/* restConfig= */ null),
+              response -> ((ErrorMessage) response.getEntity()).getErrorCode(),
+              response -> ((ErrorMessage) response.getEntity()).getMessage())
+          .build();
+
+  StreamingResponse() {
+  }
+
+  public static <T> StreamingResponse<T> from(MappingIterator<T> input) {
+    return new InputStreamingResponse<>(input);
+  }
+
+  public final <O> StreamingResponse<O> compose(
+      Function<? super T, ? extends CompletableFuture<O>> transform) {
+    return new ComposingStreamingResponse<>(this, transform);
+  }
```

Comment:
> in mind that we don't need them separately
I don't see a way of combining them, actually. They are separate because we control the calls to `next()` which are synchronous, and call `transform`, which is async, on each result individually. Having the callers call `next()` on `MappingIterator` defeats some of the purpose of this class, which is to orchestrate the input and output of data. Callers should only care about transforming input to output. We could have a `from` that takes both the `input` and the `transform`, but we would not be gaining anything, really, since internally it would be the same. I think this way reads more naturally on the calling code, as it resembles what the Java `Stream` API offers.

Reply:
I reserve my right to noodle on this once we get some breathing room :D